firmware/application/io_file.cpp

```diff
@@ -19,20 +19,76 @@
  * Boston, MA 02110-1301, USA.
  */
 
+#include "string.h"
 #include "io_file.hpp"
 
+namespace fs = std::filesystem;
+static const fs::path c8_ext = u".C8";
+
+// convert c16 buffer to c8 buffer
+// (output buffer size is bytes/2)
+void FileWriter::c16_to_c8(const void* buffer, File::Size bytes) {
+    int16_t* src = (int16_t*)buffer;
+    uint16_t* dest = (uint16_t*)buffer;
+
+    for (File::Size i = 0; i < bytes / 4; i++) {
+        uint8_t re = *src++ / 256;
+        uint8_t im = *src++ / 256;
+        *dest++ = ((uint16_t)im << 8) + re;
+    }
+}
+
+// convert c8 buffer to c16 buffer
+// (output buffer size is 2*bytes)
+void FileReader::c8_to_c16(const void* buffer, File::Size bytes) {
+    int8_t* src = (int8_t*)buffer + bytes - 1;
+    int32_t* dest = (int32_t*)((uint8_t*)buffer + 2 * bytes - 4);
+
+    for (File::Size i = 0; i < bytes / 2; i++) {
+        int16_t im = *src-- * 256;
+        int16_t re = *src-- * 256;
+        *dest-- = ((int32_t)im << 16) + re;
+    }
+}
+
+Optional<File::Error> FileReader::open(const std::filesystem::path& filename) {
+    convert_c8 = path_iequal(filename.extension(), c8_ext);  // automatically enable C8/C16 conversion based on file name
+    return file_.open(filename);
+}
+
+Optional<File::Error> FileReader::open(const std::filesystem::path& filename, bool do_c8_conversion) {
+    convert_c8 = do_c8_conversion;
+    return file_.open(filename);
+}
+
+// if C8 conversion enabled, half the number of bytes are read from the file & expanded to fill the whole buffer
 File::Result<File::Size> FileReader::read(void* const buffer, const File::Size bytes) {
-    auto read_result = file_.read(buffer, bytes);
+    auto read_result = file_.read(buffer, convert_c8 ? bytes / 2 : bytes);
     if (read_result.is_ok()) {
         bytes_read_ += read_result.value();
+
+        if (convert_c8) {
```

Comment:
QQ: Why is the conversion needed at all? Would it be possible to have the Replay proc just do the right thing while reading?

Reply:
For Replay (not for Capture), that is a very good point since I since discovered that Replay is converting the C16 data back to C8.  proc_replay is also duplicating each complex number 8 times so Replay's internal buffer is actually 8X larger than the disk buffer.  On Replay operations it would certainly make more sense to tell Replay to skip the C16->C8 conversion step if it's a C8 file and only do the 8X copy thing.

Reply:
In the C8 Replay case, the conversion could be optimized out, since it's really doing a C8->C16->C8, but I'd like to defer this optimization for a subsequent PR.  In the C8 *Capture* case or in the C32 Replay case, it's not so easy because all the baseband filters would need to be modified.